File: dyros_robot_menagerie/src/fr3_xls/robot_data.cpp
#include "fr3_xls/robot_data.h"

namespace FR3XLS
{
    static RobotData::Mobile::KinematicParam makeMobileParam()
    {
        RobotData::Mobile::KinematicParam p;
        p.type = RobotData::Mobile::DriveType::Mecanum;
        p.wheel_radius = 0.120;
        p.base2wheel_positions = {Vector2d( 0.2225,  0.2045),  // front_left
                                  Vector2d( 0.2225, -0.2045),  // front_right
                                  Vector2d(-0.2225,  0.2045),  // rear_left
                                  Vector2d(-0.2225, -0.2045)}; // rear_right
        p.base2wheel_angles = {0, 0, 0, 0};
        p.roller_angles = {-M_PI/4,  // front_left
                            M_PI/4,  // front_right
                            M_PI/4,  // rear_left
                           -M_PI/4}; // rear_right
        return p;
    }

    static RobotData::MobileManipulator::JointIndex makeJointIndex()
    {
        RobotData::MobileManipulator::JointIndex j;
        j.virtual_start = 0;
        j.mani_start = VIRTUAL_DOF;
        j.mobi_start = VIRTUAL_DOF + MANI_DOF;
        return j;
    }

    static RobotData::MobileManipulator::ActuatorIndex makeActuatorIndex()
    {
        RobotData::MobileManipulator::ActuatorIndex a;
        a.mani_start = 0;
        a.mobi_start = MANI_DOF;
        return a;
    }

    FR3XLSRobotData::FR3XLSRobotData()
    : RobotData::MobileManipulator::MobileManipulatorBase(
        makeMobileParam(),
        ament_index_cpp::get_package_share_directory("dyros_robot_menagerie") + "/robot/fr3_xls.urdf",
        ament_index_cpp::get_package_share_directory("dyros_robot_menagerie") + "/robot/fr3_xls.srdf",
        ament_index_cpp::get_package_share_directory("mujoco_ros_sim"),
        makeJointIndex(),
        makeActuatorIndex())
    {
        ee_name_ = "fr3_link8";
    }

    Affine3d FR3XLSRobotData::computePose(const VectorXd& q_virtual,
                                            const VectorXd& q_mobile,
                                            const VectorXd& q_mani)
    {
        return MobileManipulatorBase::computePose(q_virtual,q_mobile,q_mani,ee_name_);
    }

    MatrixXd FR3XLSRobotData::computeJacobian(const VectorXd& q_virtual,
                                                const VectorXd& q_mobile,
                                                const VectorXd& q_mani)
    {
        return MobileManipulatorBase::computeJacobian(q_virtual,q_mobile,q_mani,ee_name_);
    }
    MatrixXd FR3XLSRobotData::computeJacobianTimeVariation(const VectorXd& q_virtual,
                                                             const VectorXd& q_mobile,
                                                             const VectorXd& q_mani,
                                                             const VectorXd& qdot_virtual,
                                                             const VectorXd& qdot_mobile,
                                                             const VectorXd& qdot_mani)
    {
        return MobileManipulatorBase::computeJacobianTimeVariation(q_virtual,q_mobile,q_mani,qdot_virtual,qdot_mobile,qdot_mani,ee_name_);
    }

    VectorXd FR3XLSRobotData::computeVelocity(const VectorXd& q_virtual,
                                                const VectorXd& q_mobile,
                                                const VectorXd& q_mani,
                                                const VectorXd& qdot_virtual,
                                                const VectorXd& qdot_mobile,
                                                const VectorXd& qdot_mani)
    {
        return MobileManipulatorBase::computeVelocity(q_virtual,q_mobile,q_mani,qdot_virtual,qdot_mobile,qdot_mani,ee_name_);
    }

    MatrixXd FR3XLSRobotData::computeJacobianActuated(const VectorXd& q_virtual,
                                                        const VectorXd& q_mobile,
                                                        const VectorXd& q_mani)
    {
        return MobileManipulatorBase::computeJacobianActuated(q_virtual,q_mobile,q_mani,ee_name_);
    }

    MatrixXd FR3XLSRobotData::computeJacobianTimeVariationActuated(const VectorXd& q_virtual,
                                                                     const VectorXd& q_mobile,
                                                                     const VectorXd& q_mani,
                                                                     const VectorXd& qdot_virtual,
                                                                     const VectorXd& qdot_mobile,
                                                                     const VectorXd& qdot_mani)
    {
        return MobileManipulatorBase::computeJacobianTimeVariationActuated(q_virtual,q_mobile,q_mani,qdot_virtual,qdot_mobile,qdot_mani,ee_name_);
    }

    RobotData::Manipulator::ManipulabilityResult FR3XLSRobotData::computeManipulability(const VectorXd& q_mani, 
                                                                                          const VectorXd& qdot_mani, 
                                                                                          const bool& with_grad, 
                                                                                          const bool& with_graddot)
    {
        return MobileManipulatorBase::computeManipulability(q_mani,qdot_mani,with_grad,with_graddot,ee_name_);
    }

    Affine3d FR3XLSRobotData::getPose() const
    {
        return MobileManipulatorBase::getPose(ee_name_);
    }

    MatrixXd FR3XLSRobotData::getJacobian()
    {
        return MobileManipulatorBase::getJacobian(ee_name_);
    }

    MatrixXd FR3XLSRobotData::getJacobianTimeVariation()
    {
        return MobileManipulatorBase::getJacobianTimeVariation(ee_name_);
    } 

    VectorXd FR3XLSRobotData::getVelocity()
    {
        return MobileManipulatorBase::getVelocity(ee_name_);
    }   

    RobotData::Manipulator::ManipulabilityResult FR3XLSRobotData::getManipulability(const bool& with_grad, const bool& with_graddot)
    {
        return MobileManipulatorBase::getManipulability(with_grad,with_graddot,ee_name_);
    }
    
    MatrixXd FR3XLSRobotData::getJacobianActuated()
    {
        return MobileManipulatorBase::getJacobianActuated(ee_name_);
    }

    MatrixXd FR3XLSRobotData::getJacobianActuatedTimeVariation()
    {
        return MobileManipulatorBase::getJacobianActuatedTimeVariation(ee_name_);
    }
}

File: dyros_robot_menagerie/dyros_robot_menagerie/fr3_xls/robot_data.py
from pathlib import Path
from typing import Final, Tuple
import numpy as np
from ament_index_python.packages import get_package_share_directory
from dyros_robot_controller import (DriveType,
                                    KinematicParam,
                                    JointIndex,
                                    ActuatorIndex,
                                    )
from dyros_robot_controller.robot_data import MobileManipulatorBase

TASK_DOF:     Final[int] = 6
VIRTUAL_DOF:  Final[int] = 3
MANI_DOF:     Final[int] = 7
MOBI_DOF:     Final[int] = 4
ACTUATOR_DOF: Final[int] = MANI_DOF + MOBI_DOF
JOINT_DOF:    Final[int] = ACTUATOR_DOF + VIRTUAL_DOF

"""

"""
class FR3XLSRobotData(MobileManipulatorBase):
    def __init__(self) -> None:
        robot_pkg = get_package_share_directory("dyros_robot_menagerie")
        mujoco_pkg = get_package_share_directory("mujoco_ros_sim")
        
        urdf  = str(Path(robot_pkg, "robot", "fr3_xls.urdf"))
        srdf  = str(Path(robot_pkg, "robot", "fr3_xls.srdf"))
        
        mobile_kine = KinematicParam(type         = DriveType.Mecanum,
                                     wheel_radius = 0.120,
                                     base2wheel_positions = [np.array([ 0.2225,  0.2045]),
                                                             np.array([ 0.2225, -0.2045]),
                                                             np.array([-0.2225,  0.2045]),
                                                             np.array([-0.2225, -0.2045])],
                                       base2wheel_angles = [0,0,0,0],
                                       roller_angles = [-np.pi/4, 
                                                         np.pi/4,
                                                         np.pi/4,
                                                        -np.pi/4]
                                     )
        joint_idx = JointIndex(virtual_start = 0,
                               mani_start    = VIRTUAL_DOF,
                               mobi_start    = VIRTUAL_DOF + MANI_DOF,
                               )
        actuator_idx = ActuatorIndex(mani_start = 0,
                                     mobi_start = MANI_DOF,
                                     )
        
        super().__init__(mobile_param  = mobile_kine,
                         urdf_path     = urdf,
                         srdf_path     = srdf,
                         packages_path = mujoco_pkg,
                         joint_idx     = joint_idx,
                         actuator_idx  = actuator_idx,
                         )
        
        self._ee_name: Final[str] = "fr3_link8"

    def compute_pose(self, q_virtual: np.ndarray, q_mobile: np.ndarray, q_mani: np.ndarray) -> np.ndarray:
        return super().compute_pose(q_virtual, q_mobile, q_mani, self.ee_name)

    def compute_jacobian(self, q_virtual: np.ndarray, q_mobile: np.ndarray, q_mani: np.ndarray) -> np.ndarray:
        return super().compute_jacobian(q_virtual, q_mobile, q_mani, self.ee_name)

    def compute_jacobian_time_variation(self,
                                        q_virtual: np.ndarray,
                                        q_mobile: np.ndarray,
                                        q_mani: np.ndarray,
                                        qdot_virtual: np.ndarray,
                                        qdot_mobile: np.ndarray,
                                        qdot_mani: np.ndarray,
    ) -> np.ndarray:
        return super().compute_jacobian_time_variation(q_virtual, q_mobile, q_mani, qdot_virtual, qdot_mobile, qdot_mani, self.ee_name)

    def compute_velocity(self,
                        q_virtual: np.ndarray,
                        q_mobile: np.ndarray,
                        q_mani: np.ndarray,
                        qdot_virtual: np.ndarray,
                        qdot_mobile: np.ndarray,
                        qdot_mani: np.ndarray,
    ) -> np.ndarray:
        return super().compute_velocity( q_virtual, q_mobile, q_mani, qdot_virtual, qdot_mobile, qdot_mani, self.ee_name)

    def compute_jacobian_actuated(self, q_virtual: np.ndarray, q_mobile: np.ndarray, q_mani: np.ndarray) -> np.ndarray:
        return super().compute_jacobian_actuated(q_virtual, q_mobile, q_mani, self.ee_name)

    def compute_jacobian_time_variation_actuated(self,
                                                 q_virtual: np.ndarray,
                                                 q_mobile: np.ndarray,
                                                 q_mani: np.ndarray,
                                                 qdot_virtual: np.ndarray,
                                                 qdot_mobile: np.ndarray,
                                                 qdot_mani: np.ndarray,
                                                 ) -> np.ndarray:
        return super().compute_jacobian_time_variation_actuated(q_virtual, q_mobile, q_mani, qdot_virtual, qdot_mobile, qdot_mani, self.ee_name)

    def compute_manipulability(self, q_mani: np.ndarray, qdot_mani: np.ndarray, with_grad: bool = False, with_graddot: bool = False) -> Tuple[float, np.ndarray, np.ndarray]:
        return super().compute_manipulability(q_mani, qdot_mani, with_grad, with_graddot, self.ee_name)

    def get_pose(self) -> np.ndarray:
        return np.asarray(super().get_pose(self.ee_name))

    def get_jacobian(self) -> np.ndarray:
        return np.asarray(super().get_jacobian(self.ee_name))

    def get_jacobian_time_variation(self) -> np.ndarray:
        return np.asarray(super().get_jacobian_time_variation(self.ee_name))

    def get_velocity(self) -> np.ndarray:
        return np.asarray(super().get_velocity(self.ee_name))

    def get_manipulability(self, with_grad: bool = False, with_graddot: bool = False)-> Tuple[float, np.ndarray, np.ndarray]:
        return super().get_manipulability(with_grad, with_graddot, self.ee_name)

    def get_jacobian_actuated(self) -> np.ndarray:
        return np.asarray(super().get_jacobian_actuated(self.ee_name))

    def get_jacobian_actuated_time_variation(self) -> np.ndarray:
        return np.asarray(super().get_jacobian_actuated_time_variation(self.ee_name))
    
    @property
    def ee_name(self) -> str:
        return self._ee_name